src/algorithm/minkowski/gjk/mod.rs

```diff
@@ -292,35 +398,168 @@ where
         TR: Transform<P>,
         SP: SimplexProcessor<Point = P>,
     {
+        use CollisionStrategy::*;
         let mut contacts = Vec::default();
         for &(ref left_primitive, ref left_local_transform) in left.iter() {
             let left_transform = left_transform.concat(left_local_transform);
             for &(ref right_primitive, ref right_local_transform) in right.iter() {
                 let right_transform = right_transform.concat(right_local_transform);
-                match self.intersect(
+                if let Some(contact) = self.intersection(
+                    strategy,
                     left_primitive,
                     &left_transform,
                     right_primitive,
                     &right_transform,
                 ) {
-                    None => (),
-                    Some(mut simplex) => {
-                        match *strategy {
-                            CollisionStrategy::CollisionOnly => {
-                                return Some(Contact::new(CollisionStrategy::CollisionOnly));
+                    match *strategy {
+                        CollisionOnly => {
+                            return Some(contact);
+                        }
+                        FullResolution => contacts.push(contact),
+                    }
+                }
+            }
+        }
+
+        // CollisionOnly handling will have returned already if there was a contact, so this
+        // scenario will only happen when we have a contact in FullResolution mode
+        if contacts.len() > 0 {
+            // penetration depth defaults to 0., and can't be nan from EPA,
+            // so unwrapping is safe
+            contacts.into_iter().max_by(|l, r| {
```

Comment:
nit: `max_by` returns an `Option`, so no need to check `contacts.len() > 0` above